src/shackle_pool.erl

```diff
@@ -51,22 +56,89 @@ start(Name, Client, ClientOptions, Options) ->
 
 stop(Name) ->
     case options(Name) of
-        {ok, #pool_options{pool_size = PoolSize} = OptionsRec} ->
-            stop_children(server_names(Name, PoolSize)),
-            cleanup(Name, OptionsRec),
+        {ok, #pool_options{} = OptionsRec} ->
+            gen_server:call(?MODULE, {stop_pool, Name, OptionsRec}),
             ok;
         {error, Reason} ->
             {error, Reason}
     end.
 
-%% internal
--spec init() ->
-    ok.
+-spec start_link() -> {ok, pid()}.
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
 
-init() ->
+-spec init(any()) -> {ok, state()}.
+init(_) ->
     init_ets(),
     foil:new(?MODULE),
-    foil:load(?MODULE).
+    foil:load(?MODULE),
+    process_flag(trap_exit, true),
+    {ok, maps:new()}.
+
+-spec handle_call(any(), pid(), state()) -> {reply, any(), state()}.
+handle_call({stop_pool, Name, #pool_options{pool_size = PoolSize} = OptionsRec}, _From, WorkersMap) ->
+    Workers = server_names(Name, PoolSize),
+    WorkersMap2 =
+        lists:foldl(fun(Worker, Acc) ->
+            case whereis(Worker) of
+                undefined ->
+                    Acc;
+                Pid ->
+                    case maps:is_key(Pid, Acc) of
+                        true ->
+                            erlang:unlink(Pid),
+                            exit(Pid, kill),
+                            maps:remove(Pid, Acc);
+                        _ ->
+                            exit(Pid, kill),
```

Comment:
Why would some Pid is not a key of Acc? When a worker starts, its Pid got recorded in the gen_server state. And when it crashes, the Pid will be replaced with a new one. WorkersMap should have the accurate worker info, right?

Reply:
just error handling.